File: backend/src/main/java/com/fitmart/backend/service/ProductService.java
package com.fitmart.backend.service;

import com.fitmart.backend.dto.request.ProductRequest;
import com.fitmart.backend.dto.response.ProductResponse;
import org.springframework.data.domain.Page;

import java.math.BigDecimal;
import java.util.List;

public interface ProductService {

    ProductResponse create(ProductRequest request);

    ProductResponse update(
            Long id,
            ProductRequest request);

    void delete(Long id);

    ProductResponse getById(Long id);

    List<ProductResponse> getAll();

    // Search
    List<ProductResponse> search(String keyword);

    // Category Filter
    List<ProductResponse> getByCategory(Long categoryId);

    // Price Filter
    List<ProductResponse> getByPriceRange(
            BigDecimal min,
            BigDecimal max);

    // Pagination + Sorting
    Page<ProductResponse> getProducts(
            int page,
            int size,
            String sortBy,
            String direction);

}

File: frontend/src/services/productService.js
import api from "./api";

const productService = {

  getAll: async () => {
    const { data } = await api.get("/products");
    console.log("API Response:", data);

    return data.data;
  },

  getById: async (id) => {
    const { data } = await api.get(`/products/${id}`);
    return data.data;
  },

  search: async (keyword) => {
    const { data } = await api.get("/products/search", {
      params: { keyword },
    });

    return data.data;
  },

  getByCategory: async (categoryId) => {
    const { data } = await api.get(
      `/products/category/${categoryId}`
    );

    return data.data;
  },

  getByPrice: async (min, max) => {
    const { data } = await api.get("/products/price", {
      params: {
        min,
        max,
      },
    });

    return data.data;
  },

  getPage: async (
    page = 0,
    size = 12,
    sortBy = "id",
    direction = "asc"
  ) => {
    const { data } = await api.get("/products/page", {
      params: {
        page,
        size,
        sortBy,
        direction,
      },
    });

    return data.data;
  },
};

export default productService;